File: 2021/08/30/add_strings.py
"""
https://leetcode.com/problems/add-strings/
Author: Andrew Jarombek
Date: 8/30/2021
"""


def add_strings(num1: str, num2: str) -> str:
    result = ""
    carry = 0

    for i in range(max(len(num1), len(num2))):
        first = int(num1[-1 - i]) if i < len(num1) else 0
        second = int(num2[-1 - i]) if i < len(num2) else 0
        val = first + second + carry

        carry = val // 10
        result = f"{val % 10}{result}"

    if carry == 1 and len(result) > 0:
        result = f"{carry}{result}"

    return result


if __name__ == '__main__':
    assert add_strings("9", "1") == "10"

File: 2023/06/23/go/add_strings.go
/**
 * https://leetcode.com/problems/add-strings/
 * Author: Andrew Jarombek
 * Date: 6/23/2023
 */

package _go

func addStrings(num1 string, num2 string) string {
	result := ""
	carry := 0

	for i, j := len(num1)-1, len(num2)-1; i >= 0 || j >= 0 || carry > 0; i, j = i-1, j-1 {
		var digit1, digit2 int

		if i >= 0 {
			digit1 = int(num1[i] - '0')
		}

		if j >= 0 {
			digit2 = int(num2[j] - '0')
		}

		sum := digit1 + digit2 + carry
		result = string(rune(sum%10+'0')) + result
		carry = sum / 10
	}

	return result
}
